### PR TITLE
Amendment fixReducedOffersV2 fixes issue #4937

### PR DESCRIPTION
**Pull request by scottschurr**
## High Level Overview of Change

In issue https://github.com/XRPLF/rippled/issues/4937 @shortthefomo reported additional instances of order book blocking since the `fixReducedOffersV1` amendment went live.  Thanks to that report I was able to identify an additional case where blocking offers could be generated.  This pull request fixes the newly identified case.

### Context of Change

Having order books occasionally blocked by offers that were modified so that their quality was worse than the originally placed quality has been a problem for a while.  Three different ways that this could occur were identified and addressed by amendment `fixReducedOffersV1`.  All three of those case were addressed by carefully looking at and fixing the numeric rounding applied during calculations.

This new amendment, `fixReducedOffersV2`, identifies and fixes an additional rounding case that was missed during the research for `fixReducedOffersV1 `.

Beyond the fix itself, there are two kinds of unit test additions in this pull request.

1. There's a new test that directly exercises the failing case both with and without the amendment.  It demonstrates that the old rounding behavior could lead to blocked order books.  It also shows that with the new rounding behavior the order books are no longer blocked for this range of cases.

2. A few pre-existing test cases failed due to the changed rounding.  Those test cases were examined and the tests patched to handle the new rounding behavior.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Tests (you added tests for code that already exists, or your new feature included in this PR)

### API Impact

None.

### CODE REVIEW

File: src/test/app/Flow_test.cpp
Comment (gregtatcam) [inline src/test/app/Flow_test.cpp:1393]: Why need to run this test with `reducedOffersV2` enabled? There is no change to `testDirectStep` with/without the amendment.

File: src/test/app/Flow_test.cpp
Comment (gregtatcam) [inline src/test/app/Flow_test.cpp:538]: Does it make sense to extend this test to show that the order book is actually blocked without the amendment?

File: src/ripple/app/paths/impl/BookStep.cpp
Comment (gregtatcam) [inline src/ripple/app/paths/impl/BookStep.cpp:673]: Would it be beneficial to add a comment here and in `limitStepOut` explaining why the choice of the rounding in each case prevents order book blocking?

File: src/test/app/ReducedOffer_test.cpp
Comment (gregtatcam) [inline src/test/app/ReducedOffer_test.cpp:786]: Shouldn't we check for the exact count of the blocking rates since the same tests are executed on every run? If this count changes in the next release then doesn't it indicate some change that needs to be investigated?

**Review by gregtatcam — APPROVED** [review]
👍 Looks good. I left a few minor comments for you to consider.

**Review by scottschurr — COMMENTED** [review]
Thanks for the review @gregtatcam.  I've addressed a couple of your comments.  There were also two that I didn't do anything with, but I left notes about why I they are the way they are.  I'm open to continuing the conversation on those if you'd like.

File: src/test/app/Flow_test.cpp
Comment (gregtatcam) [reply src/test/app/Flow_test.cpp:1416]: I think it should be removed if it doesn't accomplish anything. It just increases the tests running time without any benefit.

File: src/test/app/ReducedOffer_test.cpp
Comment (gregtatcam) [reply src/test/app/ReducedOffer_test.cpp:786]: Yeah, it's fine.

**Review by gregtatcam — APPROVED** [review]
👍

File: src/ripple/app/paths/AMMOffer.h
Comment (seelabs) [inline src/ripple/app/paths/AMMOffer.h:107]: Consider using the globally available rules rather than passing this as a parameter. (also fine as-is, but I think I'd like to get away from passing rules around as much as we do and just using the new global).

File: src/ripple/app/paths/AMMOffer.h
Comment (seelabs) [inline src/ripple/app/paths/AMMOffer.h:118]: See comment above about passing rules as a param.

File: src/ripple/protocol/Quality.h
Comment (seelabs) [inline src/ripple/protocol/Quality.h:204]: Could use a comment for how "strict" differs from the non-strict function.

File: src/ripple/protocol/Quality.h
Comment (seelabs) [inline src/ripple/protocol/Quality.h:209]: We usually don't define functions inline in the class. I know this is a member template and that gets a little tedious outside of the class. I'd prefer it was defined out of the class, but it's a minor nit.

File: src/ripple/protocol/Quality.h
Comment (seelabs) [inline src/ripple/protocol/Quality.h:213]: This looks like a copy of `ceil_in` with one function called changed. 

If `ceil_in` is being deprecated and will be replaced with `ceil_in_strict` then I actually prefer the code duplication. And I think this is the case. If this is not the case, then let's look at merging the two.

On the same topic, if the `ceil_in` is being deprecated, consider renaming `ceil_in` to `ceil_in_deprecated` or somesuch and renaming `ceil_in_strict` to `ceil_in`.

File: src/ripple/protocol/impl/Quality.cpp
Comment (seelabs) [inline src/ripple/protocol/impl/Quality.cpp:68]: I usually like passing the callback as a parameter rather than as a template. But I'm also fine with this as-is.

**Review by seelabs — COMMENTED** [review]
I think this is ready to go, but I left a couple of nits for you to consider. I'll approve it after you take a look. I'm fine with however you decide to resolve the nits (including keeping them as they are).

File: src/ripple/protocol/Quality.h
Comment (scottschurr) [reply src/ripple/protocol/Quality.h]: I'm not sure about `ceil_in` being deprecated.  But it certainly has an odd behavior regarding rounding.  So we _might_ eventually want to replace all uses of it with `ceil_in_strict`.  But I'd like to hold off on renaming either of them until we know that we're actually going to make that change (which would have to go under an amendment).  There are a few things in this code base that have been "deprecated" for quite a few years without ever being removed.  I'd like to not add to that confusion.

**Review by seelabs — APPROVED** [review]
:+1:

**Comment by scottschurr** [comment]
Proposed commit message:
```
fixReducedOffersV2: prevent offers from blocking order books:

Fixes issue #4937.

The fxiReducedOffersV1 amendment fixed certain forms of offer
modification that could lead to blocked order books.  Reduced
offers can block order books if the effective quality of the
reduced offer is worse than the quality of the original offer
(from the perspective of the taker). It turns out that, for
small values, the quality of the reduced offer can be
significantly affected by the rounding mode used during
scaling computations.

Issue #4937 identified an additional code path that modified
offers in a way that could lead to blocked order books.  This
commit changes the rounding in that newly located code path so
the quality of the modified offer is never worse than the
quality of the offer as it was originally placed.

It is possible that additional ways of producing blocking
offers will come to light.  Therefore there be a future need
for a V3 amendment.
```

**Pull request merged**
